designs/data_serialization_design.md

```diff
@@ -0,0 +1,189 @@
+# FAIR Data Serialization Design
+## Overview
+
+This document outlines the design for serializing FAIR’s core data structures, which is essential for enabling FAIR as a distributed system. It also covers how serialized data is expected to be consumed and synchronized across hosts. 
+
+This document doesn't capture details about how the serialized states from other hosts are merged/considered in the decision making.
+
+## Glossary
+
+What's a Bucket? \
+The most basic data-structure in FAIR implementatin is a bucket - this contains a probability and a last-seen time. The probability is decayed (or) improved based on some factors. For the life-time of the Structure, a client - identified by the client-ID, consistently lands on a few buckets (based on seed). The final decision is made on top of the bucket probabilities.
```

Comment:
I'd say the probability is increased or decreased based on success/failure and it also decays over time by a configurable factor.

Reply:
Other than that, great explanation!